File: themes/lib/targets/wallpaper.py
"""Wallpaper via swww, with optional lutgen filtering and output caching."""

from __future__ import annotations

import hashlib
import os
import shutil
import subprocess
import sys
from pathlib import Path

from themes.lib.schema import ColorScheme, ThemeState

TARGET_NAME = "wallpaper"
ASSEMBLY = "command"
SYNC_SAFE = False
_CACHE_VERSION = "lutgen-apply-v1"


def _swww_command(path: str) -> list[str]:
    return [
        "swww", "img", path,
        "--transition-type", "fade",
        "--transition-duration", "1",
    ]


def _cache_root() -> Path:
    base = os.environ.get("XDG_CACHE_HOME")
    if base:
        return Path(base).expanduser() / "apply-theme" / "wallpaper"
    return Path("~/.cache/apply-theme/wallpaper").expanduser()


def _cache_key(colors: ColorScheme, wallpaper: Path) -> str:
    stat = wallpaper.stat()
    digest = hashlib.sha256()
    digest.update(_CACHE_VERSION.encode())
    digest.update(str(wallpaper.resolve()).encode())
    digest.update(str(stat.st_size).encode())
    digest.update(str(stat.st_mtime_ns).encode())
    for color in colors.palette:
        digest.update(color.lower().encode())
    return digest.hexdigest()[:16]


def _filtered_wallpaper_path(colors: ColorScheme, state: ThemeState) -> Path:
    wallpaper = Path(state.wallpaper).expanduser()
    stem = "".join(ch if ch.isalnum() or ch in ("-", "_") else "-" for ch in wallpaper.stem)
    stem = stem.strip("-_") or "wallpaper"
    scheme = "".join(ch if ch.isalnum() or ch in ("-", "_") else "-" for ch in state.color_scheme)
    return _cache_root() / f"{stem}-{scheme}-{_cache_key(colors, wallpaper)}.png"


def _run_command(cmd: list[str]) -> tuple[bool, str]:
    try:
        subprocess.run(cmd, check=True, capture_output=True, text=True)
        return True, ""
    except FileNotFoundError:
        return False, f"{cmd[0]!r} not found"
    except subprocess.CalledProcessError as exc:
        message = exc.stderr.strip() or exc.stdout.strip() or str(exc)
        return False, message


def _warn(message: str) -> None:
    print(f"  wallpaper warning: {message}", file=sys.stderr)


def _apply_wallpaper(path: Path) -> None:
    ok, message = _run_command(_swww_command(str(path)))
    if not ok:
        _warn(f"failed to apply wallpaper {path}: {message}")


def _fallback_to_source(source: Path, reason: str) -> None:
    _warn(f"{reason}; using original wallpaper")
    _apply_wallpaper(source)


def generate(colors: ColorScheme, state: ThemeState) -> list[list[str]]:
    del colors
    if state.filter_wallpaper:
        return []
    return [_swww_command(state.wallpaper)]


def on_apply(colors: ColorScheme, state: ThemeState) -> None:
    if not state.filter_wallpaper:
        return

    source = Path(state.wallpaper).expanduser()
    if not source.is_file():
        _warn(f"source wallpaper does not exist: {source}")
        return

    try:
        filtered = _filtered_wallpaper_path(colors, state)
        if not filtered.is_file():
            if shutil.which("lutgen") is None:
                _fallback_to_source(
                    source,
                    "filter_wallpaper is enabled but 'lutgen' is not installed",
                )
                return

            try:
                filtered.parent.mkdir(parents=True, exist_ok=True)
            except OSError as exc:
                _fallback_to_source(
                    source,
                    f"could not create wallpaper filter cache at {filtered.parent}: {exc}",
                )
                return

            ok, message = _run_command(
                [
                    "lutgen",
                    "apply",
                    "--cache",
                    "-o",
                    str(filtered),
                    str(source),
                    "--",
                    *colors.palette,
                ]
            )
            if not ok:
                filtered.unlink(missing_ok=True)
                _fallback_to_source(
                    source,
                    f"could not generate filtered wallpaper with lutgen: {message}",
                )
                return
        _apply_wallpaper(filtered)
    except Exception as exc:
        _fallback_to_source(
            source,
            f"unexpected error while preparing filtered wallpaper: {exc}",
        )

File: desktopctl/src/theme/targets/wallpaper.rs
use super::{Assembly, GeneratedContent, TargetMetadata};
use crate::theme::{
    expand_user_path, find_command, fnv1a_fingerprint,
    schema::{ColorScheme, ThemeState},
};
use std::{
    fs, io,
    path::{Path, PathBuf},
    process::Command,
    time::UNIX_EPOCH,
};

pub const METADATA: TargetMetadata = TargetMetadata::new(
    "wallpaper",
    Assembly::Command,
    &["color_scheme", "wallpaper", "filter_wallpaper"],
)
.sync_safe(false);

// v2: cache key digest switched from SHA-256 to the shared FNV-1a fingerprint;
// v1 entries are simply never matched again and regenerate on demand.
const CACHE_VERSION: &str = "lutgen-apply-v2";

fn awww_command(path: &str) -> Vec<String> {
    vec![
        "awww".to_owned(),
        "img".to_owned(),
        path.to_owned(),
        "--transition-type".to_owned(),
        "fade".to_owned(),
        "--transition-duration".to_owned(),
        "1".to_owned(),
    ]
}

fn cache_root() -> crate::Result<PathBuf> {
    match std::env::var("XDG_CACHE_HOME") {
        Ok(base) => Ok(PathBuf::from(base).join("apply-theme").join("wallpaper")),
        Err(std::env::VarError::NotPresent) => expand_user_path("~/.cache/apply-theme/wallpaper"),
        Err(error) => Err(io::Error::new(io::ErrorKind::InvalidInput, error).into()),
    }
}

fn cache_key(colors: &ColorScheme, wallpaper: &Path) -> crate::Result<String> {
    let stat = wallpaper.metadata()?;
    let resolved = wallpaper.canonicalize()?;
    let modified_ns = stat
        .modified()?
        .duration_since(UNIX_EPOCH)
        .map_err(io::Error::other)?
        .as_nanos()
        .to_string();

    let mut input = String::new();
    input.push_str(CACHE_VERSION);
    input.push_str(&resolved.to_string_lossy());
    input.push_str(&stat.len().to_string());
    input.push_str(&modified_ns);
    for color in &colors.palette {
        input.push_str(&color.to_ascii_lowercase());
    }
    Ok(fnv1a_fingerprint(&input))
}

pub(crate) fn filtered_wallpaper_path(
    colors: &ColorScheme,
    state: &ThemeState,
) -> crate::Result<PathBuf> {
    let wallpaper = PathBuf::from(&state.wallpaper).expanduser()?;
    let stem = sanitize_stem(
        wallpaper
            .file_stem()
            .and_then(|stem| stem.to_str())
            .unwrap_or_default(),
    );
    let scheme = sanitize_component(&state.color_scheme);
    Ok(cache_root()?.join(format!(
        "{stem}-{scheme}-{}.png",
        cache_key(colors, &wallpaper)?
    )))
}

pub(crate) fn selected_wallpaper_path(
    colors: &ColorScheme,
    state: &ThemeState,
) -> crate::Result<PathBuf> {
    let source = PathBuf::from(&state.wallpaper).expanduser()?;
    if !state.filter_wallpaper {
        return Ok(source);
    }

    match filtered_wallpaper_path(colors, state) {
        Ok(filtered) if filtered.is_file() => Ok(filtered),
        _ => Ok(source),
    }
}

fn sanitize_component(value: &str) -> String {
    value
        .chars()
        .map(|character| {
            if character.is_ascii_alphanumeric() || matches!(character, '-' | '_') {
                character
            } else {
                '-'
            }
        })
        .collect()
}

fn sanitize_stem(value: &str) -> String {
    let sanitized = sanitize_component(value);
    let trimmed = sanitized.trim_matches(['-', '_']);
    if trimmed.is_empty() {
        "wallpaper".to_owned()
    } else {
        trimmed.to_owned()
    }
}

fn run_command(command: &[String]) -> (bool, String) {
    let output = Command::new(&command[0]).args(&command[1..]).output();
    match output {
        Ok(output) if output.status.success() => (true, String::new()),
        Ok(output) => {
            let stderr = String::from_utf8_lossy(&output.stderr).trim().to_owned();
            let stdout = String::from_utf8_lossy(&output.stdout).trim().to_owned();
            let message = if !stderr.is_empty() {
                stderr
            } else if !stdout.is_empty() {
                stdout
            } else {
                output.status.to_string()
            };
            (false, message)
        }
        Err(error) if error.kind() == io::ErrorKind::NotFound => {
            (false, format!("{:?} not found", command[0]))
        }
        Err(error) => (false, error.to_string()),
    }
}

fn warn(message: &str) {
    eprintln!("  wallpaper warning: {message}");
}

fn apply_wallpaper(path: &Path) {
    let (ok, message) = run_command(&awww_command(&path.display().to_string()));
    if !ok {
        warn(&format!(
            "failed to apply wallpaper {}: {message}",
            path.display()
        ));
    }
}

fn fallback_to_source(source: &Path, reason: &str) {
    warn(&format!("{reason}; using original wallpaper"));
    apply_wallpaper(source);
}

pub fn generate(_colors: &ColorScheme, state: &ThemeState) -> crate::Result<GeneratedContent> {
    let commands = if state.filter_wallpaper {
        Vec::new()
    } else {
        vec![awww_command(&state.wallpaper)]
    };
    Ok(GeneratedContent::commands(commands))
}

pub fn on_apply(colors: &ColorScheme, state: &ThemeState) -> crate::Result<()> {
    if !state.filter_wallpaper {
        return Ok(());
    }

    let source = PathBuf::from(&state.wallpaper).expanduser()?;
    if !source.is_file() {
        warn(&format!(
            "source wallpaper does not exist: {}",
            source.display()
        ));
        return Ok(());
    }

    match filtered_wallpaper_path(colors, state) {
        Ok(filtered) => {
            if !filtered.is_file() {
                if find_command("lutgen").is_none() {
                    fallback_to_source(
                        &source,
                        "filter_wallpaper is enabled but 'lutgen' is not installed",
                    );
                    return Ok(());
                }

                if let Err(error) =
                    fs::create_dir_all(filtered.parent().expect("filtered wallpaper parent"))
                {
                    fallback_to_source(
                        &source,
                        &format!(
                            "could not create wallpaper filter cache at {}: {error}",
                            filtered
                                .parent()
                                .expect("filtered wallpaper parent")
                                .display()
                        ),
                    );
                    return Ok(());
                }

                let (ok, message) = run_command(&{
                    let mut command = vec![
                        "lutgen".to_owned(),
                        "apply".to_owned(),
                        "--cache".to_owned(),
                        "-o".to_owned(),
                        filtered.display().to_string(),
                        source.display().to_string(),
                        "--".to_owned(),
                    ];
                    command.extend(colors.palette.iter().cloned());
                    command
                });
                if !ok {
                    let _ = fs::remove_file(&filtered);
                    fallback_to_source(
                        &source,
                        &format!("could not generate filtered wallpaper with lutgen: {message}"),
                    );
                    return Ok(());
                }
            }
            apply_wallpaper(&filtered);
        }
        Err(error) => fallback_to_source(
            &source,
            &format!("unexpected error while preparing filtered wallpaper: {error}"),
        ),
    }

    Ok(())
}

trait ExpandUserPath {
    fn expanduser(self) -> crate::Result<PathBuf>;
}

impl ExpandUserPath for PathBuf {
    fn expanduser(self) -> crate::Result<PathBuf> {
        let text = self.to_string_lossy().to_string();
        expand_user_path(&text)
    }
}
